### ISSUE TITLE
Long-term support of this project

### ISSUE THREAD

**Issue by karamanolev** [fake-name/ChromeController#1]
Hello,

I stumbled upon this in HackerNews and it's exactly what I was looking for in terms of functionality, but do you plan to support this long-term? Is it safe for this to be plugged in a CI suite and can I count on it being continually worked on vs. just a random hobby project that isn't going to be touched?

**Comment by fake-name** [comment]
It'll probably be reasonably stable, once some of the outstanding bugs that're blocking me from using it at all are fixed.

Basically, right now, I have a [fairly large automatic system](https://github.com/fake-name/AutoTriever) running myself, using phantomjs. I'd *really* like to switch to using chromium for basically everything, but the inability to manipulate the browser cookies (see [issue here](https://bugs.chromium.org/p/chromium/issues/detail?id=672744)) is preventing me from switching.

Once that is solved, I'd probably put it into more maintenance mode (not a lot of API changes, but try to prevent actual breakage). I'm not sure if I'd actively try to follow chromium HEAD wrt upstream changes, though I assume that the remote debugger interface is used in enough places that major breaking changes there would be avoided unless absolutely required.

Ideally, I don't *think* any changes would need to be made even if upstream changed (just update the various protocol definition changes). 

Basically, I think it's probably safe to assume it'll be at least bugfixed for the foreseeable future, though things like py3.6 support depend on me switching to 3.6 (or contributors). I don't know if 3.6 has any breaking AST changes that would require modification.

**Comment by karamanolev** [comment]
@fake-name Any reason why you chose AST generation instead of regular code generation? I think it's going to be so much easier and cheaper if there's just a package, that contains generated code and a script to re-generate that code from the JSON describing the DevTools API? That being your library, this is just an opinion/suggestion. If you're open to such discussions, I guess they can happen here or over another channel.

Btw, I'm in the exact same boat - I used to use PhantomJS, but started finding it horrible. I just switched to xvfb + chrome, but would love to switch to --headless.

**Comment by fake-name** [comment]
> Any reason why you chose AST generation instead of regular code generation?

Mostly curiosity. And I wanted to see if I could make it work. From the README:

> Automatic wrapper class creation for the remote interface by parsing the chrome protocol.json file, and dynamic code generation through dynamic AST building. While this is not the most maintainable design, I chose it mostly because I wanted an excuse to learn/experiment with python AST manipulation.

So it was mostly a intellectual exercise. FWIW, it's completely possible to shove the generated AST through [`astor`](https://github.com/berkerpeksag/astor) and have it output python code (this is actually a large part of how I debug things)!

**Comment by karamanolev** [comment]
@fake-name As for the cookie issue - along with the test suite that you serve, can't you also monkey-patch in a page that deletes all the cookies as a super-ugly workaround?

**Comment by fake-name** [comment]
> can't you also monkey-patch in a page that deletes all the cookies as a super-ugly workaround?

That's.... not a bad idea, actually.

I don't really have a test suite so-to-speak at the moment. I've been more or less just doing hand-testing of the various functions that I personally need.  I don't actually need much of the API for my personal use, I basically just need to set up the browser context (set custom a custom user-agent, configure the cookies in the browser), navigate to a URL, retreive the page source, and perhaps periodically wait for a page to finish loading (I.e. get through cloudflare).

Principally, my use-case is large-scale automated web-scraping. I really only need a real browser to the extent that a lot of the internet is broken without a "real" browser.

**Comment by fake-name** [comment]
Oh, since you said you're currently using chromium + xvfb, I'd assume you're using selenium to run chromium? If that's true, you can *probably* just switch without needing to change library or anything. 

Basically, selenium uses webdriver internally to run chromium, and webdriver interacts with chromium via the debugger interface. You should be able to specify the binary to selenium to have it execute the headless version of chromium instead of the normal version. It should then just work.

This project itself is more intended to replace that entire pile of intermediate cruft. Basically, the webdriver protocol makes a *lot* of assumptions about what you're doing, and has **lots** of stipulations about how you're supposed to implement things. For example, you can't get the HTTP response code (the official suggested workaround is run the entire browser through a proxy just to get the response codes, which is a spectacularly stupid limitation).

The chromium remote debugger protocol has a lot more features that are not exposed through the intermediate webdriver layer, which is what motivated my creation of this library.

Anyways, the point is, if you're OK with being tied to selenium, you should be able to get headless chromium going without any substantial changes.

**Comment by karamanolev** [comment]
@fake-name Yes and no - Selenium uses WebDriver (which is just a protocol specification) through ChromeDriver, which is the thing that ties the WebDriver protocol to the DevTools protocol. But ChromeDriver also does other things with windows, tabs and whatnot, and so insists on launching it's own windowful Chrome/Chromium. It doesn't (yet) work with --headless and I didn't find relevant tickets mentioning someone is actively working on it. Also, ChromeDriver/Chrome have this stupid limitation that if you do a screenshot you only capture the visible area, so I'm reduced to launching Chrome with a 2000px high window. You also run into other weird issues that you have to work around. So I believe using a --headless Chrome/Chromium using the plain DevTools API (or even implement (a subset/superset of) the WebDriver protocol that does saner things straight with the DevTools API is going to be so much better.

**Comment by fake-name** [comment]
While I'm waiting for chromium to build *again*, I had a look at python 3.6 support, and it wound up being a [single-line fix](https://github.com/fake-name/ChromeController/commit/1594b3f437863cf23bf5f427d2692a6c989c3cbe#diff-9c6627f703b188d1643c3c8b47a44b2fL378), so I guess ChromeController now supports python 3.6 too!

**Comment by karamanolev** [comment]
Nice, thanks!

**Comment by fake-name** [comment]
Yeah, I'm hoping to get headless chromium into at least testing this weekend.

**Comment by fake-name** [comment]
So, I *just* pushed my use-case for this into production, so I'm now dependent on keeping it maintained, as well as beating the heck out of it in a real-world application.

I'm therefore going to close this issue.
